### PR TITLE
TL/NCCL: add more collectives

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
## What
Adding allgather, allgatherv, allreduce and broadcast for nccl tl

### CODE REVIEW

File: src/components/cl/basic/cl_basic_coll.c [hunk 1]
```diff
@@ -7,26 +7,43 @@
...
-                 (args->args.src.info_v.mem_type == UCC_MEMORY_TYPE_CUDA))) {
+            if ((args->dst.info_v.mem_type == UCC_MEMORY_TYPE_CUDA) &&
+                (UCC_IS_INPLACE(*args) ||
+                 (args->src.info_v.mem_type == UCC_MEMORY_TYPE_CUDA))) {
+                *tl_team = cl_basic_team->tl_nccl_team;
+                return UCC_OK;
+            }
+            break;
+        case UCC_COLL_TYPE_BCAST:
+            if (args->src.info.mem_type == UCC_MEMORY_TYPE_CUDA) {
```
Comment (bureddy) [inline src/components/cl/basic/cl_basic_coll.c:46]: @vspetrov @manjugv its not for this PR but is it make sense to build some static priority table per memory type and collective to select the team

File: src/components/cl/basic/cl_basic_coll.c [hunk 2]
```diff
@@ -7,26 +7,43 @@
...
             }
             break;
         case UCC_COLL_TYPE_ALLTOALLV:
-            if ((args->args.dst.info_v.mem_type == UCC_MEMORY_TYPE_CUDA) &&
-                (UCC_IS_INPLACE(args->args) ||
-                 (args->args.src.info_v.mem_type == UCC_MEMORY_TYPE_CUDA))) {
+            if ((args->dst.info_v.mem_type == UCC_MEMORY_TYPE_CUDA) &&
+                (UCC_IS_INPLACE(*args) ||
+                 (args->src.info_v.mem_type == UCC_MEMORY_TYPE_CUDA))) {
+                *tl_team = cl_basic_team->tl_nccl_team;
```
Comment (bureddy) [inline src/components/cl/basic/cl_basic_coll.c:41]: should default team nccl (or) ucp for alltoallv and alltoall cuda?

File: src/components/tl/nccl/tl_nccl_coll.c [hunk 3]
```diff
@@ -101,7 +112,7 @@ ucc_status_t ucc_nccl_alltoall_start(ucc_coll_task_t *coll_task)
     return status;
 }
 
-ucc_status_t ucc_tl_nccl_alltoall_init(ucc_tl_nccl_task_t *task)
+ucc_status_t ucc_nccl_alltoall_init(ucc_tl_nccl_task_t *task)
```
Comment (bureddy) [inline src/components/tl/nccl/tl_nccl_coll.c:115]: It seems UCP team followed "`ucc_tl_<team>_xx`". any reason for this change?

File: src/components/cl/basic/cl_basic_coll.c (hunk 2 above)
Comment (Sergei-Lebedev) [reply src/components/cl/basic/cl_basic_coll.c:41]: I think we need to add more fine grained control over selection here, probably priority based per collective with user control priorities.

File: src/components/tl/nccl/tl_nccl_coll.c (hunk 3 above)
Comment (Sergei-Lebedev) [reply src/components/tl/nccl/tl_nccl_coll.c:115]: I can rename it back to ucc_tl_, but shorter function names when there is no confusion looks better I think.

Comment (bureddy) [reply src/components/tl/nccl/tl_nccl_coll.c:115]: IMHO, either add _tl here (or) remove _tl from ucp func . @vspetrov  WDYT?

File: src/components/cl/basic/cl_basic_coll.c (hunk 2 above)
Comment (bureddy) [reply src/components/cl/basic/cl_basic_coll.c:41]: I think this can be another in PR

File: src/components/tl/nccl/tl_nccl_coll.c (hunk 3 above)
Comment (vspetrov) [reply src/components/tl/nccl/tl_nccl_coll.c:115]: @Sergei-Lebedev agree with @bureddy the main thing to be consistent. I personally would just keep the way it is: ucc_tl_ ucc_cl_, etc.

Comment (Sergei-Lebedev) [reply src/components/tl/nccl/tl_nccl_coll.c]: fixed

**Review by bureddy — APPROVED** [review]

**Review by manjugv — APPROVED** [review]

**Pull request merged**